### PR TITLE
COTR-61 Loading animation

### PR DESCRIPTION
**Pull request by disusen**
Implemented a loading scene that appears when switching to a new resource intensive scene.

### CODE REVIEW

File: Scenes/UI/Menus/Loading/LoadingScreen.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                case ResourceLoader.ThreadLoadStatus.Failed:
+                    GD.PrintErr("Resource loading failed.");
+                    return;
+
+                case ResourceLoader.ThreadLoadStatus.InvalidResource:
+                    GD.PrintErr("Cannot load, invalid resource.");
+                    return;
+
+                default:
+                GD.PrintErr("State doesn't exist.");
```
Comment (Elijuszek) [inline Scenes/UI/Menus/Loading/LoadingScreen.cs]: Fix indentation.

File: Scenes/Globals/Globals.cs [hunk 2]
```diff
@@ -119,4 +121,15 @@ public static float GetRandomFloat(float min, float max)
         double scaled = (sample * range) + min;
         return (float)scaled;
     }
+
+    /// <summary>
+    /// Loads loading screen scene which is used to load a new scene and deletes itself after it's done.
+    /// </summary>/// <param name="scenePath">Path to the next scene</param>
```
Comment (Karolinskis) [inline Scenes/Globals/Globals.cs:127]: Fix this summary

**Review by Karolinskis — DISMISSED** [review]

File: Scenes/UI/Menus/Loading/LoadingScreen.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        nextScene = scenePath;
+        animationPlayer.Play("TransIn");
+    }
+
+    /// <summary>
+    /// Loads new scene bit by bit, allowing the use of a loading bar to show progress.
+    /// </summary>
+    public void InitializeResourceLoader()
+    {
+        ResourceLoader.LoadThreadedRequest(nextScene);  // Begin loading
```
Comment (Elijuszek) [inline Scenes/UI/Menus/Loading/LoadingScreen.cs:42]: Add this peace of code code after this line:
`ResourceLoader.ThreadLoadStatus sceneLoadStatus = ResourceLoader.LoadThreadedGetStatus(nextScene);`

File: Scenes/UI/Menus/Loading/LoadingScreen.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        while (sceneLoadStatus != ResourceLoader.ThreadLoadStatus.Loaded)
+        {
+            // Updates the loading bar progress.
+            ResourceLoader.ThreadLoadStatus sceneLoadStatus = ResourceLoader.LoadThreadedGetStatus(nextScene);
```
Comment (Elijuszek) [inline Scenes/UI/Menus/Loading/LoadingScreen.cs:46]: Change this line to:
`sceneLoadStatus = ResourceLoader.LoadThreadedGetStatus(nextScene);`

File: Scenes/Globals/Globals.cs [hunk 5]
```diff
@@ -119,4 +121,16 @@ public static float GetRandomFloat(float min, float max)
...
+    /// Loads loading screen scene which is used to load a new scene and deletes itself after it's done.
+    /// </summary>
+    /// <param name="scenePath">Path to the next scene</param>
+    public void ChangeScene(string scenePath)
+    {
+        // Loading screen is loaded.
+        LoadingScreen loadingScreen = loadingScreenScene.Instantiate<LoadingScreen>();
+        AddChild(loadingScreen);
+        loadingScreen.LoadNewScene(scenePath);
+	}
```
Comment (Elijuszek) [inline Scenes/Globals/Globals.cs:135]: Fix indentation.

**Review by Elijuszek — CHANGES_REQUESTED** [review]
- Project doesn't even compile
- Indentation problems

File: Scenes/Globals/Globals.cs [hunk 6]
```diff
@@ -119,4 +121,16 @@ public static float GetRandomFloat(float min, float max)
...
+    /// Loads loading screen scene which is used to load a new scene and deletes itself after it's done.
+    /// </summary>
+    /// <param name="scenePath">Path to the next scene</param>
+    public void ChangeScene(string scenePath)
+    {
+        // Loading screen is loaded.
+        LoadingScreen loadingScreen = loadingScreenScene.Instantiate<LoadingScreen>();
+        AddChild(loadingScreen);
+        loadingScreen.LoadNewScene(scenePath);
+        }
```
Comment (Elijuszek) [inline Scenes/Globals/Globals.cs:135]: Inconsistent indentation

File: Scenes/UI/Menus/Loading/LoadingScreen.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        control = GetNode<Control>("CanvasLayer/Control/");
+        loadingBar = GetNode<ProgressBar>("CanvasLayer/Control/TextureRect/CenterContainer/PanelContainer/VBoxContainer/ProgressBar");
+        animationPlayer = GetNode<AnimationPlayer>("CanvasLayer/Control/LoadingAnimation");
+    }
+
+    /// <summary>
+    /// Called every frame. Used for updating fake loading bar.
+    /// </summary>
+    /// <param name="delta">Loading bar progress.</param>
+    public override void _Process(double delta)
```
Comment (Elijuszek) [inline Scenes/UI/Menus/Loading/LoadingScreen.cs:22]: _Process method is to performance heavy for just changing `loadingBar.Value`, use Tweens in _Ready method or change it inside InitializeResourceLoader method (While loop).

You can find more info about tweens [here](https://docs.godotengine.org/en/stable/classes/class_tween.html)

File: Scenes/UI/Menus/Loading/LoadingScreen.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            // Updates the loading bar progress.
+            sceneLoadStatus = ResourceLoader.LoadThreadedGetStatus(nextScene);
+            // Loads the new scene once everything has been loaded.
+            switch (sceneLoadStatus)
+            {
+                case ResourceLoader.ThreadLoadStatus.InProgress:
+                break;
+
+                case ResourceLoader.ThreadLoadStatus.Loaded:
+                    var loadedScene = ResourceLoader.LoadThreadedGet(nextScene) as PackedScene;
```
Comment (Elijuszek) [inline Scenes/UI/Menus/Loading/LoadingScreen.cs:55]: Avoid using var. Type of this variable is PackedScene

File: Scenes/UI/Menus/Loading/LoadingScreen.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                    var newRootNode = loadedScene.Instantiate();
```
Comment (Elijuszek) [inline Scenes/UI/Menus/Loading/LoadingScreen.cs:56]: Avoid using var. Type of this variable is Node

**Review by Elijuszek — CHANGES_REQUESTED** [review]

**Review by Elijuszek — CHANGES_REQUESTED** [review]
Minor comment changes are still needed, everything else seems alright.
I'm happy you decided to use Tweens to show loading progress.

**Review by Elijuszek — APPROVED** [review]
LGTM!

**Pull request merged**